src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+            string? longPath;
+            switch (header._typeFlag)
+            {
+                case TarEntryType.ExtendedAttributes or TarEntryType.GlobalExtendedAttributes:
+                    Debug.Assert(header._name != null);
+                    header._extendedAttributes = await ReadExtendedAttributesBlockAsync(archiveStream, header._typeFlag, header._size, header._name, cancellationToken).ConfigureAwait(false);
+                    break;
 
-                    Debug.Assert(_format is TarEntryFormat.Ustar or TarEntryFormat.Pax or TarEntryFormat.Gnu);
-                    if (_format == TarEntryFormat.Ustar)
+                case TarEntryType.LongLink:
+                    longPath = await ReadGnuLongPathDataBlockAsync(archiveStream, header._typeFlag, header._size, cancellationToken).ConfigureAwait(false);
+                    if (longPath != null)
                     {
-                        ReadUstarAttributes(buffer);
+                        header._linkName = longPath;
                     }
-                    else if (_format == TarEntryFormat.Gnu)
+                    break;
+
+                case TarEntryType.LongPath:
+                    longPath = await ReadGnuLongPathDataBlockAsync(archiveStream, header._typeFlag, header._size, cancellationToken).ConfigureAwait(false);
+                    if (longPath != null)
                     {
-                        ReadGnuAttributes(buffer);
+                        header._name = longPath;
                     }
-                    // In PAX, there is nothing to read in this section (empty space)
-                }
+                    break;
+
+                case TarEntryType.BlockDevice:
+                case TarEntryType.CharacterDevice:
+                case TarEntryType.Directory:
+                case TarEntryType.Fifo:
+                case TarEntryType.HardLink:
+                case TarEntryType.SymbolicLink:
+                    // No data section
+                    break;
 
-                ProcessDataBlock(archiveStream, copyData);
+                case TarEntryType.RegularFile:
+                case TarEntryType.V7RegularFile: // Treated as regular file
+                case TarEntryType.ContiguousFile: // Treated as regular file
+                case TarEntryType.DirectoryList: // Contains the list of filesystem entries in the data section
+                case TarEntryType.MultiVolume: // Contains portion of a file
+                case TarEntryType.RenamedOrSymlinked: // Might contain data
+                case TarEntryType.SparseFile: // Contains portion of a file
+                case TarEntryType.TapeVolume: // Might contain data
+                default: // Unrecognized entry types could potentially have a data section
+                    header._dataStream = await GetDataStreamAsync(archiveStream, copyData, header._size, cancellationToken).ConfigureAwait(false);
+                    if (header._dataStream is SeekableSubReadStream)
```

Comment:
do we need to check the type?

```suggestion
                    if (header._dataStream.CanSeek)
```

Reply:
Checking if it is SeekableSubReadStream also saves a nullcheck.

But also yes, it's necessary to check if the data stream is one of those specific two types because the GetDataStream method can also return a MemoryStream, when the underlying archive is unseekable and the user chooses to copy the data so it can be accessed later.